### PR TITLE
[WIP] Reanimate gradient scaling API with original scale update heuristic

### PR DESCRIPTION
**Pull request by mcarilli**
Also, windows memory failures responsible for the earlier reversion have been fixed.

This PR (initially) contains 2 commits:
* a revert of the revert
* all changes to implement the original Apex scale update heuristic, squashed into a single commit for easier diff review

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -1981,30 +1981,39 @@ def test_grad_scaling_builtins(self, device="cuda", dtype=torch.float):
...
         torch._amp_non_finite_check_and_unscale_(g, found_inf, inv_scale)
-        self.assertTrue(found_inf.item() == 0.0)
+        self.assertEqual(found_inf.item(), 0.0)
         self.assertTrue(torch.allclose(g, torch.ones(10, dtype=torch.float32, device="cuda"), atol=1e-7))
 
         found_inf.zero_()
         g = torch.tensor([float('inf')], dtype=dtype, device=device)
         torch._amp_non_finite_check_and_unscale_(g, found_inf, inv_scale)
-        self.assertTrue(found_inf.item() == 1.0)
+        self.assertEqual(found_inf.item(), 1.0)
```
Comment (ngimel) [inline test/test_cuda.py]: nit: no need to call `.item()` here, `assertEqual` will do it.

File: torch/cuda/amp/grad_scaler.py [hunk 2]
```diff
@@ -60,22 +60,28 @@ class GradScaler(object):
...
+        init_scale (float, optional, default=2.**16):  Initial scale factor.
+        growth_factor (float, optional, default=2.0):  Factor by which the scale is multiplied during
+            :meth:`update` if no inf/NaN gradients occur for ``growth_factor`` consecutive iterations.
+            The default value is recommended.
         backoff_factor (float, optional, default=0.5):  Factor by which the scale is multiplied during
-            :meth:`update` if inf/NaN gradients were found this iteration.  The default value is recommended.
-        enabled (bool,optional, default=True):  If ``False``, disables gradient scaling. :meth:`step` simply
+            :meth:`update` if inf/NaN gradients occur in an iteration.  The default value is recommended.
+        growth_interval (int, optional, default=2000):  Number of consecutive iterations without inf/NaN gradients
+            that must occur for the scale to be multiplied by ``growth_factor``.  The default value is recommended.
```
Comment (ngimel) [inline torch/cuda/amp/grad_scaler.py]: Remove `the default value is recommended`, duh. In other places too.

File: torch/cuda/amp/grad_scaler.py [hunk 3]
```diff
@@ -191,7 +207,8 @@ def unscale_(self, optimizer):
         if not self._enabled:
             return
 
-        assert self._scale is not None, self._scale_not_initialized_error("unscale")
+        assert self._scale is not None, self._not_initialized_error("unscale_", "_scale")
```
Comment (ngimel) [inline torch/cuda/amp/grad_scaler.py]: Nit: looks like these asserts come in pairs only, so you might as well have a single function for both of them

**Review by ngimel — COMMENTED** [review]
Looks good overall, hope it restores convergence. You should also skip failing windows test for now.
